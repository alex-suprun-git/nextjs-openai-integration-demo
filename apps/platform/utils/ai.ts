import { getLocale } from 'next-intl/server';
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import { OutputFixingParser, StructuredOutputParser } from 'langchain/output_parsers';
import { PromptTemplate } from '@langchain/core/prompts';
import { Document } from 'langchain/document';
import { loadQARefineChain } from 'langchain/chains';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import z from 'zod';

const modelParams = {
  model: 'o4-mini',
};

const parser = StructuredOutputParser.fromZodSchema(
  z.object({
    mood: z.string().describe('the mood of the person who wrote the journal entry.'),
    subject: z.string().describe('the subject of the journal entry.'),
    negative: z
      .boolean()
      .describe('is the journal entry negative? (i.e. does it contain negative emotions?).'),
    summary: z.string().max(100).describe('quick summary of the entire entry.'),
    color: z
      .string()
      .describe(
        'a hexidecimal color code that represents the mood of the entry. Example #0101fe for blue representing happiness.',
      ),
    sentimentScore: z
      .number()
      .describe(
        'sentiment of the text and rated on a scale from -10 to 10, where -10 is extremely negative, 0 is neutral, and 10 is extremely positive.',
      ),
    title: z.string().describe('A concise title summarizing the journal entry.'),
  }) as any,
);

const getPrompt = async (content: string, language: UserLocale) => {
  const formatInstructions = parser.getFormatInstructions();

  const templates = {
    de: 'Analysiere den folgenden Tagebucheintrag und antworte auf Deutsch. Übersetze die Formatierungsanweisungen ins Deutsche und befolge sie unbedingt! \n{formatInstructions}\n{entry}',
    en: 'Analyze the following journal entry and respond in English. Follow the instructions and format your response to match the format instructions, no matter what! \n{formatInstructions}\n{entry}',
  };

  const prompt = new PromptTemplate({
    template: templates[language],
    inputVariables: ['entry'],
    partialVariables: { formatInstructions },
  });

  const input = await prompt.format({
    entry: content,
  });

  return input;
};

export const analyzeEntry = async (content: string) => {
  const locale = await getLocale();

  const input = await getPrompt(content, locale as UserLocale);
  const model = new ChatOpenAI({
    ...modelParams,
    apiKey: process.env.OPENAI_API_KEY,
  });
  const output = await model.invoke(input);

  try {
    return await parser.parse(output.content as string);
  } catch (e) {
    const fixParser = OutputFixingParser.fromLLM(new ChatOpenAI(modelParams), parser);
    const fix = await fixParser.parse(output.content as string);
    return fix;
  }
};

export const analysisFeedback = async (
  language: UserLocale,
  question: string,
  entries: BaseEntry[],
) => {
  if (!entries.length) {
    return undefined;
  }

  const docs = entries.map(
    (entry) =>
      new Document({
        pageContent: entry.content as string,
        metadata: { source: entry.id, date: entry.createdAt },
      }),
  );
  const model = new ChatOpenAI(modelParams);
  const chain = loadQARefineChain(model);
  const embeddings = new OpenAIEmbeddings();
  const store = await MemoryVectorStore.fromDocuments(docs, embeddings);
  const relevantDocs = await store.similaritySearch(question);

  const templates = {
    en: `Answer the following question focusing exclusively on the person's mood and feelings based on the given documents. If the question touches on any other topic or if the context is unclear, please respond with a polite request to rephrase the question. Ensure your answer is in the same language as the input. \nQuestion: {question}`,
    de: `Beantworte die folgende Frage ausschließlich im Hinblick auf die Stimmung und Gefühle der Person basierend auf den gegebenen Dokumenten. Falls die Frage ein anderes Thema betrifft oder der Kontext unklar ist, bitte höflich um eine Umformulierung der Frage. Stelle sicher, dass deine Antwort in derselben Sprache wie die Eingabe erfolgt. \nFrage: {question}`,
  };

  const promptTemplate = new PromptTemplate({
    template: templates[language],
    inputVariables: ['question'],
  });

  const input = await promptTemplate.format({
    question: question,
  });

  const res = await chain.invoke({
    input_documents: relevantDocs,
    question: input,
  });

  return res.output_text;
};
