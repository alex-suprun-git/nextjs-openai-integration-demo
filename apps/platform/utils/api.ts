export const createUrl = (path: string) => window.location.origin + path;

export const updateUserPromptUsage = async (promptContentLength: number) => {
  const res = await fetch(
    new Request('/api/user', {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ promptContentLength }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const createNewEntry = async (content: string) => {
  const res = await fetch(
    new Request('/api/journal', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

// @Deprecated
export const updateEntry = async (id: string, content: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const deleteEntry = async (id: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  return;
};

export const askQuestion = async (question: string) => {
  const res = await fetch(
    new Request('/api/question', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ question }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data;
};
