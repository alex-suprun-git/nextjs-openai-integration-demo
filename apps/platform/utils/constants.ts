export const MINIMUM_CONTENT_LENGTH = 30;
export const COOKIE_NAME = 'NEXT_LOCALE';
export const ENTRIES_BASE_PATH = '/records';
