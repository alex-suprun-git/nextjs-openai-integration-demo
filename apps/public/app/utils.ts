export const createUrl = (path: string) => window.location.origin + path;

export const getPageURL = (url: string) => {
  const [, , ...restSegments] = url.split('/');
  return restSegments.join('/');
};

export async function fetcher(_url: string) {
  const url = createUrl(_url);

  const res = await fetch(
    new Request(url, {
      method: 'GET',
    }),
  );

  if (!res.ok) return;
  return res.json() as Promise<{ isSignedIn: boolean }>;
}
