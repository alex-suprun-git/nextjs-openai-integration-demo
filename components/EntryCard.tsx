'use client';

import { IoDocumentTextOutline } from 'react-icons/io5';
import { FiEdit } from 'react-icons/fi';
import { formatDate, getExcerpt } from '@/utils/helpers';

type EntryCardProps = {
  createdAt: Date;
  updatedAt: Date;
  content: string;
  color: string;
};

const EntryCard = ({ createdAt, updatedAt, content, color }: EntryCardProps) => {
  const creationDate = formatDate(createdAt);
  const updatedDate = formatDate(updatedAt);

  return (
    <div className="card relative bg-white text-primary-content shadow-md">
      <div className="card-body">
        <div
          style={{ backgroundColor: color }}
          className="absolute right-4 top-4 ml-auto h-2 w-2 rounded"
        ></div>
        <h2 className="card-title">{getExcerpt(content)}</h2>
        <div className="flex justify-between">
          <small className="flex items-center text-[12px]">
            <IoDocumentTextOutline /> <span className="pl-1">{creationDate}</span>
          </small>
          {updatedDate !== creationDate && (
            <small className="flex items-center text-[12px]">
              <FiEdit /> <span className="pl-1">{updatedDate}</span>
            </small>
          )}
        </div>
      </div>
    </div>
  );
};
export default EntryCard;
