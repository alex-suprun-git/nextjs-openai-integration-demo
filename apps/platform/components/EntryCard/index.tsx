'use client';

import { useState, useRef, useEffect } from 'react';
import { useRouter } from 'next/navigation';
import { useLocale, useTranslations } from 'next-intl';
import { FaRegTrashAlt } from 'react-icons/fa';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { IoDocumentTextOutline } from 'react-icons/io5';
import { Loading } from '@repo/global-ui';
import { formatDate } from '@/utils/helpers';
import { deleteEntry } from '@/utils/api';
import useKeyPress from '@/hooks/useKeyPress';

type EntryCardProps = {
  id: string;
  createdAt: Date;
  updatedAt?: Date;
  title?: string;
  color: string;
};

const EntryCard = ({ id, createdAt, title, color }: EntryCardProps) => {
  const router = useRouter();
  const [isLoading, setIsLoading] = useState(false);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [showActionSheet, setShowActionSheet] = useState(false);
  const [actionSheetVisible, setActionSheetVisible] = useState(false);
  const cardRef = useRef<HTMLDivElement>(null);
  const locale = useLocale();

  const t = useTranslations('JournalList');
  const c = useTranslations('Global');
  const creationDate = formatDate(new Date(createdAt), locale as UserLocale);

  useEffect(() => {
    // Handle animation timing
    if (showActionSheet) {
      // Prevent body scrolling
      document.body.style.overflow = 'hidden';

      // Show sheet with animation
      setActionSheetVisible(true);
    } else {
      // Wait for animation to complete before hiding
      const timer = setTimeout(() => {
        setActionSheetVisible(false);
      }, 300);

      // Enable scrolling
      document.body.style.overflow = '';

      return () => clearTimeout(timer);
    }

    return () => {
      document.body.style.overflow = '';
    };
  }, [showActionSheet]);

  const deleteEntryHandler = async (e?: React.MouseEvent) => {
    e?.preventDefault();
    e?.stopPropagation();
    setIsLoading(true);
    await deleteEntry(id);
    setIsLoading(false);
    router.push('/');
    router.refresh();
  };

  const toggleActionSheet = (e: React.MouseEvent | React.TouchEvent) => {
    e.preventDefault();
    e.stopPropagation();
    setShowActionSheet(!showActionSheet);
  };

  const openDeleteModal = (e: React.MouseEvent | React.TouchEvent) => {
    e.preventDefault();
    e.stopPropagation();
    setShowActionSheet(false);
    setIsModalOpen(true);
  };

  const handleCardClick = (e: React.MouseEvent) => {
    if (isModalOpen || showActionSheet) {
      e.preventDefault();
      e.stopPropagation();
    }
  };

  const closeModal = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    setIsModalOpen(false);
  };

  const handleOnModalClicks = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
  };

  useKeyPress(() => {
    setShowActionSheet(false);
    setIsModalOpen(false);
  }, ['Escape']);

  return (
    <>
      {isLoading && <Loading fullscreen />}
      <div
        data-testid="entryCard"
        ref={cardRef}
        onClick={handleCardClick}
        className="card relative h-[128px] border border-gray-200 bg-white text-primary-content transition-all duration-200"
      >
        <div className="card-body justify-between text-black">
          <div
            style={{ backgroundColor: color }}
            className="absolute right-4 top-4 ml-auto h-2 w-2 rounded"
          ></div>
          <h2 className="card-title overflow-hidden">{title || `Memo [${creationDate}]`}</h2>
          <div className="flex justify-between">
            <div className="flex flex-col 2xl:flex-row">
              <small className="flex items-center text-[12px] 2xl:mr-3">
                <IoDocumentTextOutline /> <span className="pl-1">{creationDate}</span>
              </small>
            </div>

            <div className="relative">
              <button
                aria-label={t('card.openContextMenu')}
                data-testid="entryCard-edit-button"
                onClick={(e) => toggleActionSheet(e)}
                onTouchStart={(e) => {
                  e.preventDefault();
                }}
                onTouchEnd={(e) => {
                  e.preventDefault();
                  toggleActionSheet(e);
                }}
                className="border-1 btn btn-ghost btn-xs btn-circle border-gray-300 p-1 text-gray-600 hover:bg-gray-100 focus:outline-none"
              >
                <BsThreeDotsVertical />
              </button>
            </div>
          </div>
        </div>
      </div>

      {(showActionSheet || actionSheetVisible) && (
        <div
          data-testid="entryCard-context-menu"
          className="fixed inset-0 z-[1000] flex touch-none items-end justify-center"
          style={{
            backgroundColor: showActionSheet ? 'rgba(0, 0, 0, 0.5)' : 'rgba(0, 0, 0, 0)',
            transition: 'background-color 0.3s ease-out',
            WebkitTapHighlightColor: 'transparent',
          }}
          onClick={(e) => {
            e.preventDefault();
            if (showActionSheet) toggleActionSheet(e);
          }}
          onTouchEnd={(e) => {
            e.preventDefault();
            if (showActionSheet) toggleActionSheet(e);
          }}
        >
          <div
            className="w-full max-w-md rounded-t-xl bg-white p-4 shadow-2xl"
            style={{
              transform: showActionSheet ? 'translateY(0)' : 'translateY(100%)',
              WebkitTransform: showActionSheet ? 'translateY(0)' : 'translateY(100%)',
              transition: 'transform 0.3s ease-out, -webkit-transform 0.3s ease-out',
              touchAction: 'manipulation',
            }}
            onClick={(e) => e.stopPropagation()}
            onTouchEnd={(e) => e.stopPropagation()}
          >
            <div className="mb-4 flex justify-center">
              <div className="h-1 w-16 rounded-full bg-gray-300"></div>
            </div>

            <h3 className="mb-4 text-center text-lg font-medium text-gray-800">
              {title || `Memo [${creationDate}]`}
            </h3>

            <div className="flex flex-col gap-3">
              <button
                data-testid="entryCard-delete-button"
                onClick={(e) => openDeleteModal(e)}
                onTouchEnd={(e) => {
                  e.preventDefault();
                  openDeleteModal(e);
                }}
                className="flex w-full items-center justify-center gap-2 rounded-md bg-red-50 p-4 text-red-600 focus:outline-none active:bg-red-100"
                aria-label={c('deleteEntry.actionButton')}
                style={{ touchAction: 'manipulation' }}
              >
                <FaRegTrashAlt />
                <span>{c('deleteEntry.actionButton')}</span>
              </button>

              <button
                onClick={(e) => toggleActionSheet(e)}
                onTouchEnd={(e) => {
                  e.preventDefault();
                  toggleActionSheet(e);
                }}
                className="mt-2 flex w-full items-center justify-center rounded-md bg-gray-100 p-4 text-gray-700 focus:outline-none active:bg-gray-200"
                style={{ touchAction: 'manipulation' }}
              >
                {c('deleteEntry.cancelButton')}
              </button>
            </div>
          </div>
        </div>
      )}

      {isModalOpen && (
        <div
          data-testid="delete-entry-modal"
          className="modal-open modal cursor-default"
          onClick={closeModal}
        >
          <div
            className="modal-box relative border border-gray-200 bg-white"
            onClick={handleOnModalClicks}
          >
            <h3 className="text-lg font-bold text-gray-800">{c('deleteEntry.actionButton')}</h3>
            <p className="py-4 text-gray-600">{c('deleteEntry.confirmationMessage')}</p>
            <div className="modal-action">
              <button
                data-testid="delete-cancel-button"
                className="btn btn-outline border-gray-300 bg-white text-gray-700 hover:bg-gray-50 focus:outline-none"
                onClick={closeModal}
              >
                {c('deleteEntry.cancelButton')}
              </button>
              <button
                data-testid="delete-confirm-button"
                className="btn border-0 bg-red-800 text-white hover:bg-red-900 focus:outline-none"
                onClick={deleteEntryHandler}
              >
                {c('deleteEntry.confirmButton')}
              </button>
            </div>
          </div>
          <label className="modal-backdrop" onClick={closeModal}></label>
        </div>
      )}
    </>
  );
};

export default EntryCard;
