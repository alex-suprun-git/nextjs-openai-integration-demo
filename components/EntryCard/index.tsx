'use client';

import { FaRegTrashAlt } from 'react-icons/fa';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { IoDocumentTextOutline } from 'react-icons/io5';
import { FiEdit } from 'react-icons/fi';
import { formatDate, getExcerpt } from '@/utils/helpers';
import { useState, useRef, useEffect } from 'react';
import { deleteEntry } from '@/utils/api';
import { useRouter } from 'next/navigation';
import useKeyPress from '@/hooks/useKeyPress';

type EntryCardProps = {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  content: string;
  color: string;
};

const EntryCard = ({ id, createdAt, updatedAt, content, color }: EntryCardProps) => {
  const router = useRouter();
  const creationDate = formatDate(createdAt);
  const updatedDate = formatDate(updatedAt);
  const [isContextMenuOpen, setIsContextMenuOpen] = useState(false);
  const cardRef = useRef<HTMLDivElement>(null);

  const deleteEntryHandler = async (e: React.MouseEvent, id: string) => {
    e.preventDefault();
    await deleteEntry(id);
    router.push('/journal');
    router.refresh();
  };

  const contextMenuHandler = (e: React.MouseEvent) => {
    e.preventDefault();
    setIsContextMenuOpen(!isContextMenuOpen);
  };

  useKeyPress(() => setIsContextMenuOpen(false), ['Escape']);

  const handleOuterClick = (event: MouseEvent) => {
    if (cardRef.current && !cardRef.current.contains(event.target as HTMLElement)) {
      setIsContextMenuOpen(false);
    }
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleOuterClick);
    return () => {
      document.removeEventListener('mousedown', handleOuterClick);
    };
  }, []);

  return (
    <div
      data-testid="entryCard"
      ref={cardRef}
      onContextMenu={(e) => contextMenuHandler(e)}
      className="card relative bg-white text-primary-content shadow-md"
    >
      <div className="card-body">
        <div
          style={{ backgroundColor: color }}
          className="absolute right-4 top-4 ml-auto h-2 w-2 rounded"
        ></div>
        <h2 className="card-title overflow-hidden">{getExcerpt(content)}</h2>
        <div className="flex justify-between">
          <small className="flex items-center text-[12px]">
            <IoDocumentTextOutline /> <span className="pl-1">{creationDate}</span>
          </small>
          {updatedDate !== creationDate && (
            <small className="flex items-center text-[12px]">
              <FiEdit /> <span className="pl-1">{updatedDate}</span>
            </small>
          )}

          <button data-testid="entryCard-edit-button" onClick={(e) => contextMenuHandler(e)}>
            <BsThreeDotsVertical />
          </button>
        </div>
        {isContextMenuOpen && (
          <div
            data-testid="entryCard-context-menu"
            className="absolute left-0 top-0 z-10 flex h-full w-full items-end justify-end rounded-xl bg-slate-700 bg-opacity-55"
          >
            <button
              data-testid="entryCard-delete-button"
              className="btn h-[30%] w-[100%] items-center justify-center"
              onClick={(e) => deleteEntryHandler(e, id)}
            >
              <FaRegTrashAlt />
              <span className="ml-1 font-semibold">Delete item</span>
            </button>
          </div>
        )}
      </div>
    </div>
  );
};

export default EntryCard;
