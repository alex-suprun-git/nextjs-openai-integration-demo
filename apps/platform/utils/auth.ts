import { getServerSession } from 'next-auth';
import { authOptions } from '@/app/api/auth/[...nextauth]/auth';
import { prisma } from './db';
import { Prisma } from '@prisma/client';
import { isDynamicServerError } from 'next/dist/client/components/hooks-server-context';

export const getCurrentUser = async () => {
  try {
    const session = await getServerSession(authOptions);

    if (!session?.user?.email) {
      console.warn('User email is not available. User might be signed out.');
      return null;
    }

    const user = await prisma.user.findFirstOrThrow({
      where: {
        email: session.user.email as string,
      },
    });

    return user;
  } catch (error: any) {
    if (error instanceof Prisma.PrismaClientValidationError) {
      console.error('Prisma validation error:', error.message);
      throw new Error('Failed to find user due to validation error.');
    } else if (error.code === 'P2025') {
      console.error('User not found:', error.message);
      throw new Error('User not found.');
    } else {
      if (isDynamicServerError(error)) {
        throw error;
      }
      console.error('Unexpected error:', error);
      throw new Error('An unexpected error occurred.');
    }
  }
};
