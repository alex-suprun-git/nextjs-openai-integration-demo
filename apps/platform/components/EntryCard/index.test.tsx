import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach, Mock } from 'vitest';
import EntryCard from '../EntryCard';
import { formatDate } from '@/utils/helpers';
import { deleteEntry } from '@/utils/api';
import { useRouter } from 'next/navigation';
import { useTranslations, useLocale, IntlProvider } from 'next-intl';

vi.mock('@/utils/helpers', () => ({
  formatDate: vi.fn(),
}));
vi.mock('@/utils/api', () => ({
  deleteEntry: vi.fn(),
}));
vi.mock('next/navigation', () => ({
  useRouter: vi.fn(),
}));
vi.mock('next-intl', async (importOriginal) => {
  const actual = (await importOriginal()) as any;
  return {
    ...actual,
    useTranslations: vi.fn(),
    useLocale: vi.fn(),
    IntlProvider: ({ children }: { children: React.ReactNode }) => <>{children}</>,
  };
});

describe('EntryCard', () => {
  const mockRouterPush = vi.fn();
  const mockRouterRefresh = vi.fn();

  const mockEntry = {
    id: '1',
    createdAt: new Date('2025-01-01T00:00:00Z'),
    updatedAt: new Date('2025-01-02T00:00:00Z'),
    title: 'This is a test entry',
    color: '#ff0000',
  };

  const mockTranslations: { [key: string]: string } = {
    'JournalList.card.openContextMenu': 'Open context menu',
    'Global.deleteEntry.actionButton': 'Delete this memo',
    'Global.deleteEntry.confirmationMessage': 'Are you sure you want to delete this memo?',
    'Global.deleteEntry.cancelButton': 'Cancel',
    'Global.deleteEntry.confirmButton': 'Delete',
  };

  beforeEach(async () => {
    (useTranslations as Mock).mockImplementation((prefix: string) => {
      return (key: string) => {
        const fullKey = prefix ? `${prefix}.${key}` : key;
        return mockTranslations[fullKey] || key;
      };
    });
    (useLocale as Mock).mockReturnValue('en');
    (formatDate as Mock).mockImplementation((date: Date) => date.toISOString().split('T')[0]);
    (useRouter as Mock).mockReturnValue({
      push: mockRouterPush,
      refresh: mockRouterRefresh,
    });
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  const renderWithProvider = (component: React.ReactElement) =>
    render(
      <IntlProvider locale="en" messages={{}}>
        {component}
      </IntlProvider>,
    );

  it('renders correctly and matches snapshot', () => {
    const { container } = renderWithProvider(<EntryCard {...mockEntry} />);
    expect(container).toMatchSnapshot();
  });

  it('renders with correct text and date', () => {
    renderWithProvider(<EntryCard {...mockEntry} />);
    let titles = screen.getAllByText('This is a test entry');
    titles.forEach((title) => {
      expect(title).toBeInTheDocument();
    });

    expect(screen.getByText('2025-01-01')).toBeInTheDocument();
  });

  it('opens dropdown menu when clicking the three dots button', () => {
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Find the three dots button and click
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);

    // Check that the menu opened (action sheet in the component)
    const actionSheet = document.getElementById(`action-sheet-${mockEntry.id}`);
    expect(actionSheet).toBeInTheDocument();
    expect(actionSheet).toHaveAttribute('open');
    expect(screen.getByText('Delete this memo')).toBeInTheDocument();
  });

  it('closes dropdown when clicking outside', async () => {
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Open the dropdown menu
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);

    // Check that the menu opened
    const actionSheet = document.getElementById(`action-sheet-${mockEntry.id}`);
    expect(actionSheet).toBeInTheDocument();
    expect(actionSheet).toHaveAttribute('open');

    // Click the backdrop to close
    const backdrop = screen.getByText('close');
    fireEvent.click(backdrop);

    // Verify it's closed
    await waitFor(() => {
      expect(actionSheet).not.toHaveAttribute('open');
    });
  });

  it('opens modal when clicking delete in the dropdown', () => {
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Open the dropdown menu
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);

    // Find and click the delete button
    const deleteButton = screen.getByTestId('entryCard-delete-button');
    fireEvent.click(deleteButton);

    // Check that the modal window opened
    expect(screen.getByTestId('delete-entry-modal')).toBeInTheDocument();
    expect(screen.getByText('Are you sure you want to delete this memo?')).toBeInTheDocument();
    expect(screen.getByTestId('delete-cancel-button')).toBeInTheDocument();
    expect(screen.getByTestId('delete-confirm-button')).toBeInTheDocument();
  });

  it('closes modal when clicking cancel button', async () => {
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Open the dropdown menu and then the modal window
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);
    const deleteButton = screen.getByTestId('entryCard-delete-button');
    fireEvent.click(deleteButton);

    // Check that the modal window opened
    expect(screen.getByTestId('delete-entry-modal')).toBeInTheDocument();

    // Find and click the Cancel button
    const cancelButton = screen.getByTestId('delete-cancel-button');
    fireEvent.click(cancelButton);

    // Check that the modal window closed
    await waitFor(() => {
      expect(screen.queryByTestId('delete-entry-modal')).not.toBeInTheDocument();
    });
  });

  it('calls deleteEntry and router.push when confirming deletion in modal', async () => {
    (deleteEntry as Mock).mockResolvedValueOnce({});
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Open the dropdown menu and then the modal window
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);
    const deleteButton = screen.getByTestId('entryCard-delete-button');
    fireEvent.click(deleteButton);

    // Check that the modal window opened
    expect(screen.getByTestId('delete-entry-modal')).toBeInTheDocument();

    // Click the Delete button in the modal window
    const confirmDeleteButton = screen.getByTestId('delete-confirm-button');
    fireEvent.click(confirmDeleteButton);

    // Check that the necessary functions were called
    await waitFor(() => {
      expect(deleteEntry).toHaveBeenCalledWith(mockEntry.id);
      expect(mockRouterPush).toHaveBeenCalledWith('/');
      expect(mockRouterRefresh).toHaveBeenCalled();
    });
  });

  it('closes dropdown and modal when pressing Escape key', async () => {
    renderWithProvider(<EntryCard {...mockEntry} />);

    // Open the dropdown menu
    const dotsButton = screen.getByTestId('entryCard-edit-button');
    fireEvent.click(dotsButton);

    // Check that the menu opened
    const actionSheet = document.getElementById(`action-sheet-${mockEntry.id}`);
    expect(actionSheet).toBeInTheDocument();
    expect(actionSheet).toHaveAttribute('open');

    // Press Escape
    fireEvent.keyDown(document, { key: 'Escape' });

    // Check that the dropdown closed
    await waitFor(() => {
      expect(actionSheet).not.toHaveAttribute('open');
    });

    // Open the dropdown menu and then the modal window
    fireEvent.click(dotsButton);
    const deleteButton = screen.getByTestId('entryCard-delete-button');
    fireEvent.click(deleteButton);

    // Check that the modal window opened
    expect(screen.getByTestId('delete-entry-modal')).toBeInTheDocument();

    // Press Escape
    fireEvent.keyDown(document, { key: 'Escape' });

    // Check that the modal window closed
    await waitFor(() => {
      expect(screen.queryByTestId('delete-entry-modal')).not.toBeInTheDocument();
    });
  });
});
