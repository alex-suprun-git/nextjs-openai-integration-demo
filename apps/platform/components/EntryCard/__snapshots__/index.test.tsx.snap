// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntryCard > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="card relative h-[128px] border border-gray-200 bg-white text-primary-content transition-all duration-200"
    data-testid="entryCard"
  >
    <div
      class="card-body justify-between text-black"
    >
      <div
        class="absolute right-4 top-4 ml-auto h-2 w-2 rounded"
        style="background-color: rgb(255, 0, 0);"
      />
      <h2
        class="card-title overflow-hidden"
      >
        This is a test entry
      </h2>
      <div
        class="flex justify-between"
      >
        <div
          class="flex flex-col 2xl:flex-row"
        >
          <small
            class="flex items-center text-[12px] 2xl:mr-3"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 221.25V416a48 48 0 0 1-48 48H144a48 48 0 0 1-48-48V96a48 48 0 0 1 48-48h98.75a32 32 0 0 1 22.62 9.37l141.26 141.26a32 32 0 0 1 9.37 22.62z"
                fill="none"
                stroke-linejoin="round"
                stroke-width="32"
              />
              <path
                d="M256 56v120a32 32 0 0 0 32 32h120m-232 80h160m-160 80h160"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
            </svg>
             
            <span
              class="pl-1"
            >
              2025-01-01
            </span>
          </small>
        </div>
        <div
          class="relative"
        >
          <button
            aria-label="Open context menu"
            class="border-1 btn btn-ghost btn-xs btn-circle border-gray-300 p-1 text-gray-600 hover:bg-gray-100 focus:outline-none"
            data-testid="entryCard-edit-button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0m0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0m0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
